November 2018 activation code

sickpig opened the pull request:
Summary:

- Activation code for May 2018 has been removed completely since all the changes (blk size, new op_codes) where never used before the fork. 

- EB default value changed to 32MB, before it was changed during init step

- EB value can't be lower than 32MB, this was the case also before these set of changes. But the constraint was enforced during `UpdateTip` execution. With code in this PR you can't even set the value of this tweak lower then 32MB, cause the validator will raise an error. 

- MG (max generated block size) default value changed to 8MB. I'm open to set it back to 2MB, I chose to change it because in the may 2018 activation code the value of MG was changed to 8MB. Actually there was an a little bit of ambiguity cause client that were on during the upgrade got the value set to 8MB, if after the fork it got rebooted for whatever reason MG got it back to 2MB.

TODO:

- ~~need to fix `validateblocktemplate.py` test, I temporarily disabled a little chunk of the test due to the fact that it is not possible to set EB lower than 32MB anymore.~ (done) 

- ~~excessive.py need to be fixed~~ (done) 

- ~~need to find out block height for testnet may 2018 fork.~~

sickpig committed:
Change default EB and MG block size according to May '18 specification

Default EB is 32MB, default max generated blocksize (MG) is 8MB since the
May, 15th 2018 upgrade has been activated. See https://github.com/bitcoincashorg/bitcoincash.org/blob/master/spec/may-2018-hardfork.md
for more details.

While at it remove logic to force MG to 8MB a block before May'18 activation.

sickpig committed:
Remove activation code for May '18 re-enabled opcodes.

Re-enabled opcodes where never used before, we are hence
permitting something it wasn't before. The probability of
having a reorg that go back to may '18 is very low, therefore
we could remove the activation code.

sickpig committed:
Disabled functional tests for May, 15 '18 activation code

Keep the actual files around so that people could use it as template
to build tests for future upgrades activation code.

sickpig committed:
Add MIN_EXCESSIVE_BLOCK_SIZE and set it 32MB.

Setting EB lower than this will fork off the client from
the current BCH conensus.

sickpig committed:
Enforce 32MB as the minimum value for Excessive Block size.

Since May 2018 the maximum block size accepted by clients implementing
the Bitcoin Cash protocol is 32MB. So having an EB value less than 32
will fork you off the net work.

Enforce this constraint in the tweak validator rather than in main.cpp
and init.cpp.

sickpig committed:
Comment code to clear the mempool in case reorg to pre may '18 upgrade

The May '18 activating block has been already checkpointed and the
probability to have a reorg back to pre May '18 height would mean
that we have to face more pressing issues.

sickpig committed:
Use MIN_EXCESSIVE_BLOCK_SIZE and DEFAULT_BLOCK_MAX_SIZE rather than raw value

sickpig committed:
Temporarely disabled excessive block check in validateblocktemplate.py

Since there's no way to set an EB lower than 32MB, we need to modify the
code to generate a block bigger than EB min value and then test if the
block template validation raise an error due cause we generated a
self-orphaned block.

sickpig committed:
Introduce two new consensus paramas nov2018ActivationTime and may2018Height

Rename activation function helper accordingly, from IsMay152018Enabled
to IsNov152018Enabled. We could do this because IsMay152018Enabled is not used
anymore in the code because all May 2018 changes were strictly an addition to
the protocol. I.e. never before a a block bigger than 8MB was produced and never
before the re-enabled op_codes where used.

sickpig committed:
Renable code to clear mempool in case the nov '18 fork block get reorged

sickpig committed:
Encoforce minimum Excessive Block Size (EBS) only on the main and test net.

That will let us test for "excessive" condition in regtest mode w/o
the need to create massive blocks in size. The actual min limit for
EB on regtest is 1000 bytes.

sickpig committed:
Adapt various tests for new EB/MG defaults

sickpig committed:
Use GetConsnesus to set miningForkTime default

Rather than setting the default using a numeric literal (1542300000)
use the value defined in consensus/params.h and set in chainparams.h
(Params().GetConsenus().nov2018ActivationTime).

We needed to pass the nertowrk type to Params(), cause it is still not initialized
when we defined miningForkTime.

sickpig committed:
Check that values lower than EB min value can't be set

sickpig commented:
all tests fixed, taking out of WIP

awemany commented on qa/rpc-tests/excessive.py:
```diff
@@ -74,15 +74,6 @@ def setup_network(self, split=False):
-        # So we set the mocktime to sometime far in advance of the fork.
-        MAY152018_START_TIME = 1526400000;
-        self.nodes[0].setmocktime(MAY152018_START_TIME - 10000)
-        self.nodes[1].setmocktime(MAY152018_START_TIME - 10000)
```
A `for` loop saves two lines of code here...

sickpig replied:
how so? we removed this 6 lines.

awemany replied:
Ooops. Yeah, I guess I am colorblind :D

awemany commented on qa/rpc-tests/excessive.py:
```diff
@@ -210,14 +201,14 @@ def testExcessiveSigops(self):
-        self.nodes[3].set("net.excessiveBlock=5000000")
         self.nodes[0].setminingmaxblock(5000000)
         self.nodes[1].setminingmaxblock(5000000)
         self.nodes[2].setminingmaxblock(5000000)
```
Same comment about `for` loop here...

sickpig replied:
ok, makes sense. will do

awemany commented on src/chainparams.cpp:
```diff
@@ -133,7 +133,9 @@ class CMainParams : public CChainParams
-        consensus.may2018activationTime = 1526400000;
+        consensus.may2018Height = 530359;
+        // May, 15 2018 hard fork
+        consensus.nov2018ActivationTime = 1542300000;
```
This can be similarly factored out, please have a look at #1311.

sickpig replied:
yes we could do it, wonder if it is worth doing cause we are going to change it again just after Nov activation to transform it into an height based param. Once we do this we basically have different values for each net...

awemany replied:
I see. I was looking at the may stuff, which sat there just like that, though. No strong opinion either way.

sickpig replied:
me neither. I'm ok either way. Let's see what others think about it.

awemany commented on src/main.cpp:
```diff
@@ -839,10 +837,6 @@ static bool AcceptToMemoryPoolWorker(CTxMemPool &pool,
 
         // Set extraFlags as a set of flags that needs to be activated.
         uint32_t extraFlags = 0;
-        if (hasMay152018)
```
@gandrewstone told me to use a `CTweak` just for CDS/V - but wouldn't it make the most sense to add a SCRIPT_ENABLE_NOV2018_OPCODES flag and use the code here in the same way as for the may upgrade?

sickpig replied:
I think @gandrewstone told you to that while waiting for this PR to be finalized. I think that you can use the `IsNov152018Enabled()` helper and do it the same way we did in May.

awemany replied:
Makes sense, thanks.

sickpig committed:
Revert "Use GetConsnesus to set miningForkTime default"

This reverts commit c0611f66e83fd589a5a2c4e7673aae16b605a23f.

sickpig committed:
Use loop to change EB and MG to make the code more compact.

ptschip commented on src/main.cpp:
```diff
@@ -839,10 +837,6 @@ static bool AcceptToMemoryPoolWorker(CTxMemPool &pool,
 
         // Set extraFlags as a set of flags that needs to be activated.
         uint32_t extraFlags = 0;
```
do we still need extraflags?

sickpig replied:
As the code stands we don't. It is a left over from the op_code activation from May 2018 upgrade. Will remove. But I guess @awemany could use it again from op_checkdatasigverify.

sickpig committed:
Remove May '18 op_codes activation code

sickpig commented:
Test a full IBD from scratch with these options:

- checkpoints=0  (disable checkpoints)
- checklevel=4   (for each block `ConnectBlock` is performed)

The node with this patch applied sync to current tip correctly. It took 8 hrs 20 min.

ptschip commented:
needs rebase

sickpig commented:
PR rebased after #1340 went in. Automatic merge wasn't able to solve a conflict in `main.cpp`.

Last time to "solve" is to find out the block height of last testnet fork. It is not necessary to make the code works but I think it is a good thing to have it.

sickpig commented:
Probably travis would file due to #1348.

sickpig committed:
Add testnet May, 18 fork block height to chainparams.cpp

sickpig commented:
Added testnet block height for May' 18 fork.

gandrewstone commented on src/globals.cpp:
```diff
@@ -266,20 +266,21 @@ CTweakRef<unsigned int> maxDataCarrierTweak("mining.dataCarrierSize",
-    32000000); // May2018 HF proposed max block size
+    // May2018 HF proposed max block size, no need to change it for the November '18 since we are not going to have an
+    // blopck size increase
+    MIN_EXCESSIVE_BLOCK_SIZE);
```
Since the block size is not changing, I do not think we need miningForkEB and miningForkMG

sickpig replied:
done

gandrewstone commented on src/chainparams.cpp:
```diff
@@ -427,8 +434,10 @@ class CRegTestParams : public CChainParams
         consensus.daaHeight = 0;
+        // May, 15 2018 hard fork is always 0 on regtest
+        consensus.may2018Height = 0;
         // May, 15 2018 hard fork
```
comment change to Nov

sickpig replied:
done

gandrewstone commented on src/chainparams.cpp:
```diff
@@ -133,7 +133,9 @@ class CMainParams : public CChainParams
         // May, 15 2018 hard fork
-        consensus.may2018activationTime = 1526400000;
+        consensus.may2018Height = 530359;
+        // May, 15 2018 hard fork
```
change comment to Nov

sickpig replied:
done

gandrewstone commented on src/chainparams.cpp:
```diff
@@ -259,7 +261,9 @@ class CUnlParams : public CChainParams
         // May, 15 2018 hard fork
-        consensus.may2018activationTime = 1526400000;
+        consensus.may2018Height = 0;
+        // May, 15 2018 hard fork
```
change comment to Nov

sickpig replied:
done

gandrewstone commented on src/test/excessiveblock_test.cpp:
```diff
@@ -20,29 +22,52 @@ BOOST_AUTO_TEST_CASE(rpc_excessive)
 
-    BOOST_CHECK_NO_THROW(CallRPC("setminingmaxblock 2000"));
+    // Testing the parsing of input parameters of setexcessive block,
+    // this RPC set the value for EB and AD and expect exaclty 2 unsigned
```
spelling of exactly

sickpig replied:
done

gandrewstone commented on src/test/excessiveblock_test.cpp:
```diff
@@ -184,49 +209,46 @@ BOOST_AUTO_TEST_CASE(check_validator_rule)
+    // The first execution check the validity of the value we want to set the param to.
+    // The second execution happens only if the assignment happened succesfully and
+    // could be used as notification update mechanism.
+    // Tthe first time with the 3rd par (validate) = true and the 2n with validate = false.
```
typo: Tthe and 2n -> 2nd

sickpig replied:
done

sickpig committed:
Fix comments thinkos and spelling errors

sickpig committed:
Remove miningForkEB and miningForkMG

Those are not used in the upcoming Nov '18 protocol upgrades,
they could be reintroduced if needed in the future.
